File: src/core/state_machine.cpp
/**
 * @file state_machine.cpp
 * @brief System state machine implementation
 * @version 2.0.0
 * @date 25 April 2026
 *
 * State flow: INIT → SELECT_CONDITION → SIMULATING ↔ PAUSED
 */

#include "core/state_machine.h"

// ============================================================================
// CONSTRUCTOR
// ============================================================================
StateMachine::StateMachine() {
    currentState = SystemState::INIT;
    selectedCondition = 0;  // Default: Normal
    currentEditMode = UIEditMode::CONDITION_SELECT;
    onStateChange = nullptr;
}

// ============================================================================
// PROCESS EVENTS
// ============================================================================
void StateMachine::processEvent(SystemEvent event, uint8_t param) {
    SystemState oldState = currentState;
    SystemState newState = currentState;

    switch (currentState) {
        case SystemState::INIT:
            if (event == SystemEvent::INIT_COMPLETE) {
                newState = SystemState::SELECT_CONDITION;
            }
            break;

        case SystemState::SELECT_CONDITION:
            switch (event) {
                case SystemEvent::SELECT_CONDITION:
                    selectedCondition = param;
                    break;
                case SystemEvent::START_SIMULATION:
                    newState = SystemState::SIMULATING;
                    break;
                case SystemEvent::BTN_MODE_PRESS:
                    // In condition select, mode starts simulation
                    newState = SystemState::SIMULATING;
                    break;
                case SystemEvent::BTN_UP_PRESS:
                    // Next condition
                    selectedCondition = (selectedCondition + 1) % (uint8_t)PPGCondition::COUNT;
                    break;
                case SystemEvent::BTN_DOWN_PRESS:
                    // Previous condition
                    if (selectedCondition == 0)
                        selectedCondition = (uint8_t)PPGCondition::COUNT - 1;
                    else
                        selectedCondition--;
                    break;
                default:
                    break;
            }
            break;

        case SystemState::SIMULATING:
            switch (event) {
                case SystemEvent::BTN_MODE_PRESS:
                    // Cycle edit mode
                    currentEditMode = (UIEditMode)(((uint8_t)currentEditMode + 1) % (uint8_t)UIEditMode::COUNT);
                    break;
                case SystemEvent::PAUSE:
                    newState = SystemState::PAUSED;
                    break;
                case SystemEvent::STOP:
                    newState = SystemState::SELECT_CONDITION;
                    currentEditMode = UIEditMode::CONDITION_SELECT;
                    break;
                default:
                    break;
            }
            break;

        case SystemState::PAUSED:
            switch (event) {
                case SystemEvent::RESUME:
                    newState = SystemState::SIMULATING;
                    break;
                case SystemEvent::BTN_MODE_PRESS:
                    // Resume on mode press
                    newState = SystemState::SIMULATING;
                    break;
                case SystemEvent::STOP:
                    newState = SystemState::SELECT_CONDITION;
                    currentEditMode = UIEditMode::CONDITION_SELECT;
                    break;
                default:
                    break;
            }
            break;

        case SystemState::ERROR:
            if (event == SystemEvent::INIT_COMPLETE) {
                newState = SystemState::SELECT_CONDITION;
            }
            break;
    }

    // Notify state change
    if (newState != oldState) {
        currentState = newState;
        if (onStateChange != nullptr) {
            onStateChange(oldState, newState);
        }
    }
}

// ============================================================================
// CALLBACK
// ============================================================================
void StateMachine::setStateChangeCallback(void (*callback)(SystemState, SystemState)) {
    onStateChange = callback;
}

// ============================================================================
// STRING CONVERSION
// ============================================================================
const char* StateMachine::stateToString(SystemState state) {
    switch (state) {
        case SystemState::INIT:              return "INIT";
        case SystemState::SELECT_CONDITION:  return "SELECT_CONDITION";
        case SystemState::SIMULATING:        return "SIMULATING";
        case SystemState::PAUSED:            return "PAUSED";
        case SystemState::ERROR:             return "ERROR";
        default:                             return "UNKNOWN";
    }
}

const char* StateMachine::eventToString(SystemEvent event) {
    switch (event) {
        case SystemEvent::INIT_COMPLETE:     return "INIT_COMPLETE";
        case SystemEvent::BTN_MODE_PRESS:    return "BTN_MODE";
        case SystemEvent::BTN_UP_PRESS:      return "BTN_UP";
        case SystemEvent::BTN_DOWN_PRESS:    return "BTN_DOWN";
        case SystemEvent::SELECT_CONDITION:  return "SELECT_CONDITION";
        case SystemEvent::START_SIMULATION:  return "START_SIMULATION";
        case SystemEvent::PAUSE:             return "PAUSE";
        case SystemEvent::RESUME:            return "RESUME";
        case SystemEvent::STOP:              return "STOP";
        case SystemEvent::ERROR_OCCURRED:    return "ERROR";
        default:                             return "UNKNOWN";
    }
}

const char* StateMachine::editModeToString(UIEditMode mode) {
    switch (mode) {
        case UIEditMode::CONDITION_SELECT:   return "Condition";
        case UIEditMode::EDIT_HR:            return "Heart Rate";
        case UIEditMode::EDIT_PI:            return "Perf. Index";
        case UIEditMode::EDIT_NOISE:         return "Noise";
        default:                             return "Unknown";
    }
}

File: PPG_simulator_raspi/core/state_machine.py
"""
state_machine.py — System state machine for PPG Signal Simulator.

Port of state_machine.cpp.
State flow: INIT → SELECT_CONDITION → SIMULATING ↔ PAUSED
"""

# ─── System States ───
STATE_INIT             = 0
STATE_SELECT_CONDITION = 1
STATE_SIMULATING       = 2
STATE_PAUSED           = 3
STATE_ERROR            = 4

# ─── System Events ───
EVT_INIT_COMPLETE     = 0
EVT_BTN_MODE_PRESS    = 1
EVT_BTN_UP_PRESS      = 2
EVT_BTN_DOWN_PRESS    = 3
EVT_SELECT_CONDITION  = 4
EVT_START_SIMULATION  = 5
EVT_PAUSE             = 6
EVT_RESUME            = 7
EVT_STOP              = 8
EVT_ERROR             = 9

# ─── UI Edit Modes ───
EDIT_CONDITION_SELECT = 0
EDIT_HR               = 1
EDIT_PI               = 2
EDIT_SPO2             = 3
EDIT_RR               = 4
EDIT_NOISE            = 5
EDIT_COUNT            = 6

_STATE_NAMES = {
    STATE_INIT: "INIT", STATE_SELECT_CONDITION: "SELECT_CONDITION",
    STATE_SIMULATING: "SIMULATING", STATE_PAUSED: "PAUSED", STATE_ERROR: "ERROR",
}

_EDIT_NAMES = {
    EDIT_CONDITION_SELECT: "Condition", EDIT_HR: "Heart Rate",
    EDIT_PI: "Perf. Index", EDIT_SPO2: "SpO2",
    EDIT_RR: "Resp. Rate", EDIT_NOISE: "Noise",
}

CONDITION_COUNT = 6


class StateMachine:
    """System state machine with edit mode cycling."""

    def __init__(self):
        self.state = STATE_INIT
        self.selected_condition = 0
        self.edit_mode = EDIT_CONDITION_SELECT
        self._on_state_change = None

    def set_state_change_callback(self, callback):
        """Set callback: callback(old_state, new_state)."""
        self._on_state_change = callback

    def process_event(self, event: int, param: int = 0):
        """Process a system event."""
        old = self.state
        new = old

        if old == STATE_INIT:
            if event == EVT_INIT_COMPLETE:
                new = STATE_SELECT_CONDITION

        elif old == STATE_SELECT_CONDITION:
            if event == EVT_SELECT_CONDITION:
                self.selected_condition = param
            elif event in (EVT_START_SIMULATION, EVT_BTN_MODE_PRESS):
                new = STATE_SIMULATING
            elif event == EVT_BTN_UP_PRESS:
                self.selected_condition = (self.selected_condition + 1) % CONDITION_COUNT
            elif event == EVT_BTN_DOWN_PRESS:
                self.selected_condition = (self.selected_condition - 1) % CONDITION_COUNT

        elif old == STATE_SIMULATING:
            if event == EVT_BTN_MODE_PRESS:
                self.edit_mode = (self.edit_mode + 1) % EDIT_COUNT
            elif event == EVT_PAUSE:
                new = STATE_PAUSED
            elif event == EVT_STOP:
                new = STATE_SELECT_CONDITION
                self.edit_mode = EDIT_CONDITION_SELECT

        elif old == STATE_PAUSED:
            if event in (EVT_RESUME, EVT_BTN_MODE_PRESS):
                new = STATE_SIMULATING
            elif event == EVT_STOP:
                new = STATE_SELECT_CONDITION
                self.edit_mode = EDIT_CONDITION_SELECT

        elif old == STATE_ERROR:
            if event == EVT_INIT_COMPLETE:
                new = STATE_SELECT_CONDITION

        if new != old:
            self.state = new
            if self._on_state_change:
                self._on_state_change(old, new)

    @staticmethod
    def state_to_string(state: int) -> str:
        return _STATE_NAMES.get(state, "UNKNOWN")

    @staticmethod
    def edit_mode_to_string(mode: int) -> str:
        return _EDIT_NAMES.get(mode, "Unknown")
